### PR TITLE
Upgrade write_vtu_in_parallel based on mpi large IO update

### PR DESCRIPTION
**Pull request by pengfej**
Replace the MPI I/O routines for DataOut::write_vtu_in_parallel:
- Use the large count facility to support >2GB+ individual writes
- Use write_at_all instead of write_shared (this is up to 10x faster in testing on Frontera)

### CODE REVIEW

**Comment by tjhei** [comment]
This is WIP and depends on #13611

**Comment by pengfej** [comment]
I will remove all the extra comments and space after all so don't worry!

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:7694]:
```suggestion
  const unsigned int n_ranks = Utilities::MPI::n_mpi_processes(comm);
```

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:7728]: Can you use the following instead?
```suggestion
  header_size = Utilities::MPI::broadcast(header_size, comm, 0);
```

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:7800]: Is this the right function call, or don't you need one from the big_mpi library?

Comment (pengfej) [reply source/base/data_out_base.cc:7799]: To the best of my knowledge, the header and footer are very small in terms of size, so using write_at and write_at_c (The one in big_mpi library) will not make a difference :)

Comment (tjhei) [reply source/base/data_out_base.cc:7800]: I agree with Pengfei.

File: source/base/data_out_base.cc
Comment (tjhei) [reply source/base/data_out_base.cc]: I think that function has an unfortunate design with using the return value and no real advantage for a single int over Bcast. But we can make the change if you insist.

File: source/base/data_out_base.cc
Comment (tjhei) [inline source/base/data_out_base.cc:7726]: This call needs to be outside of the if, otherwise only rank 0 will execute it but we clearly want everyone to participate.

File: source/base/data_out_base.cc
Comment (bangerth) [reply source/base/data_out_base.cc:7800]: Could we do it just out of consistency and so that people like me don't have to think about the question in the future :-)

File: source/base/data_out_base.cc
Comment (bangerth) [reply source/base/data_out_base.cc]: I find it easier to read with the return value, but no, I don't insist.

Comment (bangerth) [reply source/base/data_out_base.cc]: (Also substantially less easy to make mistakes when someone changes the code because, for example, it automatically detects the type of the object being broadcasted.)

File: source/base/data_out_base.cc
Comment (pengfej) [reply source/base/data_out_base.cc:7800]: Sure!

**Review by tjhei — APPROVED** [review]
Looks good. Can you please squash your commits?

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
Let me test this some more before merging...

**Comment by tjhei** [comment]
My suggestion would be to create a new block before ``if (myrank == 0)`` that does something like:
```
if (myrank ==0 || myrank == n_ranks -1)
{
ierr = MPI_Sendrecv_replace(&footer_size, 1, MPI_UINT64_T, n_ranks =1, 0, 0, 0, comm, MPI_STATUS_IGNORE);
}
```
You will need to delete the other ``MPI_Send`` and ``MPI_Recv`` but keep the assignment of ``footer_size`` on rank 0 (and move the declaration of this variable up and outside the ``if``).

**Comment by pengfej** [comment]
@tjhei  rank=1 tests passed, but rank=3 test freezes for a very long time, I'm not very sure if it's a code error or my machine is not working properly.

File: source/base/data_out_base.cc
Comment (tjhei) [inline source/base/data_out_base.cc:7773]:
```suggestion
    if (myrank == n_ranks - 1)
        footer_offset = size_on_proc + offset;
```

File: source/base/data_out_base.cc
Comment (tjhei) [inline source/base/data_out_base.cc:7787]: add the ``AssertThrow`` here

File: source/base/data_out_base.cc
Comment (tjhei) [reply source/base/data_out_base.cc:7773]: You never set the ``footer_offset`` correctly (you ended up creating a local variable here and setting it).

Comment (pengfej) [reply source/base/data_out_base.cc:7773]: I see. Let me changed that and test it again, thanks for pointing out!

Comment (pengfej) [reply source/base/data_out_base.cc]: The test mpirun=3 still freezes, I'm wondering if there's a way to know where exactly it freezes.

Comment (tjhei) [reply source/base/data_out_base.cc]: There is, of course. I will show you next time.

Comment (pengfej) [reply source/base/data_out_base.cc]: sounds good! I just read about the sendrecv_replace function and it seems like I might misordered the sender and the receiver, I'm testing on that fix now.

**Comment by tjhei** [comment]
Can you switch source and dest of MPI_SendRecv_replace?

**Comment by pengfej** [comment]
> 



> Can you switch source and dest of MPI_SendRecv_replace?

Just did! The tests are still forzen unfortunately..

**Comment by pengfej** [comment]
`<?xml version="1.0" ?>                                                                                                                                                                                                                       <!--                                                                                                                                                                                                                                         # vtk DataFile Version 3.0                                                                                                                                                                                                                   #This file was generated by the deal.II library on 15:30:04 at 2022/5/26                                                                                                                                                                     -->                                                                                                                                                                                                                                          <VTKFile type="UnstructuredGrid" version="0.1" compressor="vtkZLibDataCompressor" byte_order="LittleEndian">                                                                                                                                 <UnstructuredGrid>                                                                                                                                                                                                                           <Piece NumberOfPoints="4" NumberOfCells="1" >                                                                                                                                                                                                  <Points>                                                                                                                                                                                                                                       <DataArray type="Float32" NumberOfComponents="3" format="binary">                                                                                                                                                                        AQAAADAAAAAwAAAAFAAAAA==eNpjYEAGDfYMWPkgGsIGADHwAv0=                                                                                                                                                                                             </DataArray>                                                                                                                                                                                                                               </Points>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 <Cells>                                                                                                                                                                                                                                        <DataArray type="Int32" Name="connectivity" format="binary">                                                                                                                                                                             AQAAABAAAAAQAAAAEwAAAA==eNpjYGBgYARiZiBmAmIAADwABw==                                                                                                                                                                                             </DataArray>                                                                                                                                                                                                                                 <DataArray type="Int32" Name="offsets" format="binary">                                                                                                                                                                                  AQAAAAQAAAAEAAAADAAAAA==eNpjYWBgAAAAFAAF                                                                                                                                                                                                         </DataArray>                                                                                                                                                                                                                                 <DataArray type="UInt8" Name="types" format="binary">                                                                                                                                                                                    AQAAAAEAAAABAAAACQAAAA==eNrjBAAACgAK                                                                                                                                                                                                             </DataArray>                                                                                                                                                                                                                               </Cells>                                                                                                                                                                                                                                     <PointData Scalars="scalars">                                                                                                                                                                                                                  <DataArray type="Float32" Name="x" format="binary">                                                                                                                                                                                      AQAAABAAAAAQAAAADgAAAA==eNpjYGBwYEDCAAcQAQE=                                                                                                                                                                                                     </DataArray>                                                                                                                                                                                                                               </PointData>                                                                                                                                                                                                                                </Piece>                                                                                                                                                                                                                                     </UnstructuredGrid>                                                                                                                                                                                                                         </VTKFile>  `

This is the output generated after a timeout on mpirun=3 test. It seems like the only part missing is the last chunk, here's the standard output:

`
DEAL:0::hyper_cube
<?xml version="1.0" ?> 
<!-- 
# vtk DataFile Version 3.0
#This file was generated 
-->
<VTKFile type="UnstructuredGrid" version="0.1" compressor="vtkZLibDataCompressor" byte_order="LittleEndian">
<UnstructuredGrid>
<Piece NumberOfPoints="4" NumberOfCells="1" >
  <Points>
    <DataArray type="Float32" NumberOfComponents="3" format="binary">
AQAAADAAAAAwAAAAFAAAAA==eNpjYEAGDfYMWPkgGsIGADHwAv0=
    </DataArray>
  </Points>

  <Cells>
    <DataArray type="Int32" Name="connectivity" format="binary">
AQAAABAAAAAQAAAAEwAAAA==eNpjYGBgYARiZiBmAmIAADwABw==
    </DataArray>
    <DataArray type="Int32" Name="offsets" format="binary">
AQAAAAQAAAAEAAAADAAAAA==eNpjYWBgAAAAFAAF
    </DataArray>
    <DataArray type="UInt8" Name="types" format="binary">
AQAAAAEAAAABAAAACQAAAA==eNrjBAAACgAK
    </DataArray>
  </Cells>
  <PointData Scalars="scalars">
    <DataArray type="Float32" Name="x" format="binary">
AQAAABAAAAAQAAAADgAAAA==eNpjYGBwYEDCAAcQAQE=
    </DataArray>
  </PointData>
 </Piece>
 </UnstructuredGrid>
</VTKFile>

<?xml version="1.0"?>
<!--
#This file was generated 
-->
<VTKFile type="PUnstructuredGrid" version="0.1" byte_order="LittleEndian">
  <PUnstructuredGrid GhostLevel="0">
    <PPointData Scalars="scalars">
    <PDataArray type="Float32" Name="x" format="ascii"/>
    </PPointData>
    <PPoints>
      <PDataArray type="Float32" NumberOfComponents="3"/>
    </PPoints>
    <Piece Source="output.vtu"/>
  </PUnstructuredGrid>
</VTKFile>

DEAL:0::OK

DEAL:1::OK


DEAL:2::OK

`

File: source/base/data_out_base.cc
Comment (tjhei) [inline source/base/data_out_base.cc:7776]: remove this whole block.

File: source/base/data_out_base.cc
Comment (tjhei) [inline source/base/data_out_base.cc:7791]: Just write the footer on rank ``n_ranks-1`` instead of 0.

**Review by tjhei — COMMENTED** [review]
I have a simpler suggestion: Just write the footer from the last rank instead of rank 0. This way you don't need to do any communication and you can use the ``footer_offset`` direcly.

File: source/base/data_out_base.cc
Comment (pengfej) [reply source/base/data_out_base.cc]: That fixed version passed the test on my machine so I pushed it! Thank you so much!!

**Review by tjhei — APPROVED** [review]
I can confirm that things work correctly now.

**Comment by tjhei** [comment]
This is ready to go from our side. Any final comments?

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:7722]:
```suggestion
      // Write the header on rank 0 at the start of a file, i.e., offset 0.
```

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:7749]:
```suggestion
    // Use prefix sum to find specific offset to write at.
```

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:7777]:
```suggestion
        // Writing footer:
```

**Review by bangerth — APPROVED** [review]
Looks good. It tests alright, so I'll merge, but if you wouldn't mind, please make the comments I have about comments into a separate PR.

**Comment by tjhei** [comment]
@pengfej Please create a follow-up with the fixes suggested by @bangerth , ok?

**Comment by pengfej** [comment]
> @pengfej Please create a follow-up with the fixes suggested by @bangerth , ok?

 Ohh sure! Sorry I completely missed it. I only saw the merged notification.

**Pull request merged**
